core/modules/ccontrol/UserQueryDropTable.cc

```diff
@@ -0,0 +1,187 @@
+            _messageStore->addMessage(-1, 1051, message, MessageSeverity::MSG_ERROR);
+            _qState = ERROR;
+            return;
+        }
+        LOGF(getLogger(), LOG_LVL_DEBUG, "table status: %s" % statusMap[_tableName]);
+        if (statusMap[_tableName] != css::KEY_STATUS_READY) {
+            std::string message = "Unexpected status for table: " + _dbName + "."
+                            + _tableName + ": " + statusMap[_tableName];
+            _messageStore->addMessage(-1, 1051, message, MessageSeverity::MSG_ERROR);
+            _qState = ERROR;
+            return;
+        }
+    } catch (css::CssError const& exc) {
+        LOGF(getLogger(), LOG_LVL_ERROR, "css failure: %s" % exc.what());
+        std::string message = "CSS error: " + std::string(exc.what());
+        _messageStore->addMessage(-1, 1051, message, MessageSeverity::MSG_ERROR);
+        _qState = ERROR;
+        return;
+    }
+
+    // Add this query to QMeta so that progress can be tracked,
+    // QMeta needs to be updated by watcher when it finishes with the table
+    // so we embed query id into CSS table status below
+    qmeta::QInfo::QType qType = qmeta::QInfo::ASYNC;
+    std::string user = "anonymous";    // we do not have access to that info yet
+    std::string query = "DROP TABLE " + _dbName  + "." + _tableName;
+    qmeta::QInfo qInfo(qType, _qMetaCzarId, user, query, "", "", "");
+    qmeta::QMeta::TableNames tableNames;
+    qmeta::QueryId qMetaQueryId = 0;
+    try {
+        qMetaQueryId = _queryMetadata->registerQuery(qInfo, tableNames);
+    } catch (qmeta::Exception const& exc) {
+        // not fatal, just print error message and continue
+        LOGF(getLogger(), LOG_LVL_WARN, "QMeta failure (non-fatal): %s" % exc.what());
+    }
+
+    // update status to trigger watcher
+    std::string newStatus = css::KEY_STATUS_DROP_PFX + std::to_string(::time(nullptr)) +
+            ":qid=" + std::to_string(qMetaQueryId);
+    LOGF(getLogger(), LOG_LVL_DEBUG, "new table status: %s" % newStatus);
+    try {
+        // TODO: it's better to do it in one atomic operation with
+        // getTableStatus, but CSS API does not have this option yet
+        _css->setTableStatus(_dbName, _tableName, newStatus);
+        _qState = SUCCESS;
+    } catch (css::NoSuchTable const& exc) {
+        // Have it disappeared already?
+        LOGF(getLogger(), LOG_LVL_ERROR, "table disappeared from CSS");
+        std::string message = "Unknown table " + _dbName + "." + _tableName;
+        _messageStore->addMessage(-1, 1051, message, MessageSeverity::MSG_ERROR);
+        _qState = ERROR;
+    } catch (css::CssError const& exc) {
+        LOGF(getLogger(), LOG_LVL_ERROR, "CSS failure: %s" % exc.what());
+        std::string message = "CSS error: " + std::string(exc.what());
+        _messageStore->addMessage(-1, 1051, message, MessageSeverity::MSG_ERROR);
+        _qState = ERROR;
+    }
+
+    // if failed then update status in QMeta
```

Comment:
Success "happens" at a later stage when watcher actually drops table from all databases. The status of this query in CSS is updated then by watcher. At this point the query is still executing.